content/sd/0004-coordenacao-e-consenso.md

```diff
@@ -536,10 +536,339 @@ _"eventually"_) perfeito.
+processo propõe um único valor. O objetivo do algoritmo é fazer com que os processos
+corretos concordem com um vetor de valores, um para cada processo. Por exemplo, o
+objetivo poderia ser para cada processo de um grupo obter a mesma informação sobre
+os estados respectivos (de cada processo).
+
+- Conjunto de $N$ processos
+- Cada processo $p_i$ propõe um valor ($\text{input}_i$)
+- Todos os processo decidem o mesmo vetor $V$ ($\text{output}$)
+- O vetor $V$ decidido tem uma entrada por cada processo em que:
+  - ou $V[i] = \text{input}_i$
+  - ou $V[i] = null$
+
+Propriedades:
+
+1. **Terminação**: todos os processos correctos decidem ("_alguma-vez_")
+2. **Acordo uniforme**: se dois processos decidem, decidem o mesmo vetor $V$
+3. **Integridade**: se o processo $p_i$ não falhar, $V[i] = \text{input}_i$
+
+:::details[Pseudocódigo das implementações]
+Consenso usando Coerência Interativa:
+
+```php
+Quando Consenso.propoe(valor):
+  CoerenciaInteractiva.propoe(valor)
+
+Quando CoerenciaInteractiva.decide(vector):
+  valor = primeiraEntradaDiferenteDeNull(vector);
+  Consenso.decide(valor)
+
+```
+
+Coerência Interativa usando consenso:
+
+```php
+// este codigo é executado por todos os processos
+
+fun PRONTO(vector):
+  se para todo o p_x: p_x não pertence a falhados e tivermos vector_proposta[x] != null
+    retorna VERDADEIRO
+  caso contrário
+    retorna FALSO
+
+Init:
+  falhados = {}
+  para cada valor de i < N
+    vector_proposta[i] = null;
+
+Quando falha(p_x):
+  falhados = falhados U {p_x}
+
+Quando IC.propoe(valor_i)
+  DifusaoFiavel.envia(p_i, valor_i)
+
+Quando DifusaoFiavel.entrega(p_j, valor_j)
+  vector_proposta[j] = valor_j;
+
+Quando PRONTO(vector_proposta)
+  Consenso.propoe(vector_proposta) // só propõe uma vez
+
+Quando Consenso.decide(vector)
+  IC.decide(vector)
+```
+
+:::
+
+#### Derivar Consenso a partir de Coerência Interativa
+
+Por vezes é possível derivar uma solução para um problema utilizando uma solução
+para outro. Esta propriedade é muito útil porque aumenta a nossa compreensão dos
+problemas e economiza esforço de implementação.
+
+Suponha que existem as seguintes soluções para o consenso (C) e para a consistência
+interativa (IC):
+
+- $C_i(v_1, v_2, ..., v_N)$ retorna o valor de decisão do processo $p_i$ numa
+  execução da solução para o problema do consenso, onde $v_1, v_2, ..., v_N$ são
+  os valores propostos pelos processos
+- ${IC}_i(v_1, v_2, ..., v_N)[j]$ retorna o j-ésimo valor no vetor de decisão do
+  processo $p_i$ numa execução da solução para o problema da consistência interativa,
+  onde $v_1, v_2, ..., v_N$ são os valores propostos pelos processos
+
+Caso a maioria dos processos estejam corretos, construímos uma solução executando
+IC para produzir um vetor de valores em cada processo, e depois aplicando uma certa
+função sobre os valores do vetor para derivar um único valor:
+
+$$C_i(v_1, ..., v_N) = majority({IC}_i(v_1, ..., v_N)[1], ..., {IC}_i(v_1, ..., v_N)[N])$$
```

Comment:
Math blocks require that `$$` are on their own lines, otherwise they're just inline-math.

```suggestion
$$
C_i(v_1, ..., v_N) = majority({IC}_i(v_1, ..., v_N)[1], ..., {IC}_i(v_1, ..., v_N)[N])
$$
```